towhee/dag/graph_repr.py

```diff
@@ -46,12 +45,125 @@ def operators(self) -> Dict[str, OperatorRepr]:
+        Returns:
+            Return `False` if there is no loop, else `True` and the loop message.
+        """
+        # if `cur` is not the input of any operator, it will not form a loop anyway
+        if cur not in flag:
+            return False, ''
+        # If `cur` has been searched and not in a loop
+        if flag[cur] == 2:
+            return False, ''
+        # If `cur` has been visited in this search
+        if flag[cur] == 1:
+            return True, f'The dataframes {cur_list} forms a loop.'
+        flag[cur] = 1
+
+        for col in adj[cur]:
+            cur_list.append(cur)
+            status, msg = GraphRepr.dfs(col, adj, flag, cur_list)
+            if status:
+                return status, msg
+        flag[cur] = 2
+
+        return False, ''
+
+    def has_loop(self) -> Tuple[bool, str]:
+        """Check if there are loop(s) inside the graph.
+        Returns:
+            Return `False` if there is no loop, else `True` and the loop message.
+        """
+        adj = {}
+        flag = {}
+        dataframes = []
+
+        for op in self._operators:
+            out = op.outputs[0]['df']
+            for df in op.inputs:
+                name = df['df']
+                if name not in dataframes:
+                    adj[name] = []
+                    flag[name] = 0
+                    dataframes.append(name)
+                adj[name].append(out)
+
+        # Recursive DFS
+        for df in dataframes:
+            status, msg = self.dfs(df, adj, flag, [])
+            if status:
+                logging.error(msg)
+                return status, msg
+
+        return False, ''
+
+    def has_isolated_df(self):
+        non_iso_df = set()
+        iso_df = set()
+
+        for op in self._operators:
+            for i in op.inputs:
+                non_iso_df.add(i['df'])
+            non_iso_df.add(op.outputs[0]['df'])
+
+        for df in self._dataframes:
+            if df.name not in non_iso_df:
+                iso_df.add(df.name)
+
+        status = bool(iso_df)
+        msg = '' if not iso_df else f'The DAG contains isolated dataframe(s) {iso_df}.'
+        if status:
+            logging.error(msg)
+
+        return status, msg
+
+    def has_isolated_op(self):
+        if len(self._operators) == 1:
+            return False, ''
+
+        in_df = set()
+        out_df = set()
+        iso_op = set()
+
+        for op in self._operators:
+            for i in op.inputs:
+                in_df.add(i['df'])
+            out_df.add(op.outputs[0]['df'])
+
+        for op in self._operators:
+            cur_in = {i['df'] for i in op.inputs}
+            cur_out = {i['df'] for i in op.outputs}
+
+            if not cur_in.intersection(out_df - cur_out) and not cur_out.intersection(in_df - cur_in):
+                iso_op.add(op.name)
+
+        status = bool(iso_op)
+        msg = '' if not iso_op else f'The DAG contains isolated operator(s) {iso_op}.'
```

Comment:
We should either log the error in this function or return the message to the caller, but not both.

Reply:
fixed